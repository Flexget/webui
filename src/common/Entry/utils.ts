import { RawEntry, CardType, FieldNames, MappingType, GettersType } from './types';
import { Entry, DefaultEntry } from './fields';
import {
  RawMovieEntry,
  MovieEntry,
  movieFieldList,
  MovieFields,
  MovieFieldNames,
} from './fields/movies';
import {
  RawSeriesEntry,
  SeriesEntry,
  seriesFieldList,
  SeriesFields,
  SeriesFieldNames,
} from './fields/series';

import {
  RawEpisodeEntry,
  EpisodeEntry,
  EpisodeFields,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !!entry.seriesSeason;
const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<MovieFields>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<SeriesFields>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<EpisodeFields>(EpisodeFieldNames, episodesFieldList);

const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    type: CardType.Movie,
    ...getMovieProps(entry),
  };
};

const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  type: CardType.Series,
  ...getSeriesProps(entry),
});

const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  type: CardType.Episode,
  ...getEpisodeProps(entry),
});

const toDefaultEntry = (entry: RawEntry): DefaultEntry => ({
  ...entry,
  type: CardType.Default,
});

export const toEntry = (entry: RawEntry): Entry => {
  if (isMovie(entry)) {
    return toMovieEntry(entry);
  }
  if (isSeries(entry)) {
    return toSeriesEntry(entry);
  }
  if (isEpisode(entry)) {
    return toEpisodeEntry(entry);
  }
  return toDefaultEntry(entry);
};
