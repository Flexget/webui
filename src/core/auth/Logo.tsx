import React, { FC } from 'react';
import { css } from '@emotion/core';
import { Theme } from '@material-ui/core';
import headerImage from 'images/header.png';

export const logo = (theme: Theme) => css`
  background: transparent url(${headerImage}) no-repeat center;
  min-height: 9rem;
  background-size: 100% auto;
  margin: 0 1rem;
  ${theme.breakpoints.up('sm')} {
    background-size: auto;
  }
`;

const Logo: FC = () => <div css={logo} />;
export default Logo;
