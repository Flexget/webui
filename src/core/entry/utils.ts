import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
