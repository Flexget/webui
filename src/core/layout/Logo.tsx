import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import { css } from '@emotion/core';
import { Theme } from '@material-ui/core';
import headerImage from 'images/header.png';

const logo = (theme: Theme) => css`
  background: ${theme.palette.secondary.main} url(${headerImage}) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  min-height: ${theme.mixins.toolbar.minHeight};

  ${theme.breakpoints.up('sm')} {
    width: 6rem;
    background-size: 25rem;
  }
`;

const openCss = (theme: Theme) => css`
  ${theme.breakpoints.up('sm')} {
    width: 19rem;
    background-size: 17.5rem;
  }
`;

interface Props {
  sidebarOpen: boolean;
  className?: string;
}

const Logo: FC<Props> = ({ sidebarOpen, className }) => (
  <Link to="/">
    <div css={theme => [logo(theme), sidebarOpen && openCss(theme)]} className={className} />
  </Link>
);

export default Logo;
