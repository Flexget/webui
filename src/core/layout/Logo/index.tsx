import * as React from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { Link } from 'react-router-dom';
import theme from 'theme';
import headerImage from 'images/header.png';

const BaseLogo = styled.div`
  background: ${theme.palette.secondary[900]} url(${headerImage}) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  transition: ${theme.transitions.create(['width', 'background-size'])};
`;

const openCss = (open: boolean) => css`
  ${theme.breakpoints.up('sm')} {
    width: ${open ? '19rem' : '6rem'};
    background-size: ${open ? '17.5rem' : '25rem'};
  }
`;

interface Props {
  sidebarOpen: boolean;
}

const Logo: React.FC<Props> = ({ sidebarOpen }) => (
  <Link to="/">
    <BaseLogo css={openCss(sidebarOpen)} />
  </Link>
);

export default Logo;
