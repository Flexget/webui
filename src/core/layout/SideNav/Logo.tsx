import React, { FC } from 'react';
import { css } from '@emotion/core';
import { Theme } from '@material-ui/core';
import headerImage from 'images/header.png';
import { Link } from 'common/styles';

const logo = (theme: Theme) => css`
  background: ${theme.palette.secondary.main} url(${headerImage}) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  min-height: ${theme.typography.pxToRem(theme.mixins.appBar.minHeight)};

  ${theme.breakpoints.up('sm')} {
    width: ${theme.typography.pxToRem(theme.mixins.sidebar.width.closed)};
    background-size: 25rem;
  }
`;

const openCss = (theme: Theme) => css`
  width: ${theme.typography.pxToRem(theme.mixins.sidebar.width.open - 48)};
  ${theme.breakpoints.up('sm')} {
    background-size: 17.5rem;
  }
`;

interface Props {
  sidebarOpen: boolean;
  className?: string;
}

const Logo: FC<Props> = ({ sidebarOpen, className }) => (
  <Link to="/">
    <div css={theme => [logo(theme), sidebarOpen && openCss(theme)]} className={className} />
  </Link>
);

export default Logo;
