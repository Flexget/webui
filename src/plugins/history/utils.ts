import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
