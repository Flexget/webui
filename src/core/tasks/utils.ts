import { RawEntry } from 'core/entry/types';

export const toExecuteRequest = (entry?: RawEntry) => {
  if (entry) {
    const { id, title, originalUrl, ...fields } = entry;
    return { title, url: originalUrl, fields };
  }
  return { title: '', url: '', fields: {} };
};
