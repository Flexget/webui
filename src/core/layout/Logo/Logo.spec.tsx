import * as React from 'react';
import renderer from 'react-test-renderer';
import { themed, router } from 'utils/tests';
import Logo from 'core/layout/Logo';

describe('core/layout/components/Logo', () => {
  it('renders correctly with sideBarOpen', () => {
    const tree = renderer.create(router(themed(<Logo sidebarOpen />))).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('adds logoMini class without sideBarOpen', () => {
    const tree = renderer.create(router(themed(<Logo sidebarOpen={false} />))).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
