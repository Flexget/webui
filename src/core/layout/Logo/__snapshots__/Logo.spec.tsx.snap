// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/layout/components/Logo adds logoMini class without sideBarOpen 1`] = `
.emotion-0 {
  background: #263238 url(test-file-stub) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  -webkit-transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms,background-size 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms,background-size 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

<a
  href="/"
  onClick={[Function]}
>
  <div
    className="emotion-0"
  />
</a>
`;

exports[`core/layout/components/Logo renders correctly with sideBarOpen 1`] = `
.emotion-0 {
  background: #263238 url(test-file-stub) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  -webkit-transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms,background-size 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms,background-size 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

<a
  href="/"
  onClick={[Function]}
>
  <div
    className="emotion-0"
  />
</a>
`;
